### PR TITLE
[storage] Controller finalization

### PR DESCRIPTION
**Pull request by bkirwi**
- Always run shard finalization during startup. (If it's enabled!) Without this, we sometimes need to eg. drop unrelated shards to re-trigger finalization.
- Behind a new flag, force-downgrade the controller's critical handle during finalization. This is necessary to help us clean out old / bad shards from the WAL that never had their handles downgraded.

### Motivation

We're seeing some environments accumulate large numbers of shards in the finalization WAL. These should help!

### Tips for reviewer

While it would be sketchy for Persist to force-downgrade the handles, it seems less alarming for the controller to do it to its own handle. (It shouldn't put stuff in the WAL if it's still using it.)

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/storage-controller/src/command_wals.rs [hunk 1]
```diff
@@ -157,6 +157,9 @@ where
                 .await;
         }
 
+        // Finalize any other shards that have been added to the WAL.
+        self.finalize_shards().await;
```
Comment (danhhz) [inline src/storage-controller/src/command_wals.rs:161]: Looks like the other call to `finalize_shards` is protected by the `enable_storage_shard_finalization` config, so we should probably do the same here (potentially by moving the config check into the impl of `fn finalize_shards`?)

File: src/storage-types/src/dyncfgs.rs [hunk 2]
```diff
@@ -22,7 +22,19 @@ pub const DELAY_SOURCES_PAST_REHYDRATION: Config<bool> = Config::new(
         (namely, upsert) till after rehydration is finished",
 );
 
+/// Whether rendering should use `mz_join_core` rather than DD's `JoinCore::join_core`.
```
Comment (danhhz) [inline src/storage-types/src/dyncfgs.rs:25]: nit: copy-pasta on the rustdoc (and also the "This was original `false`" comment?)

File: src/storage-controller/src/lib.rs [hunk 3]
```diff
@@ -3057,11 +3058,17 @@ where
         let persist_client = &persist_client;
         let diagnostics = &Diagnostics::from_purpose("finalizing shards");
 
+        // TODO: wire up a feature flag for this!
```
Comment (danhhz) [inline src/storage-controller/src/lib.rs:3061]: did the TODO, right?

**Review by danhhz — APPROVED** [review]

**Comment by benesch** [comment]
> While it would be sketchy for Persist to force-downgrade the handles, it seems less alarming for the controller to do it to its own handle. (It shouldn't put stuff in the WAL if it's still using it.)

This makes sense to me!

File: src/storage-types/src/dyncfgs.rs (hunk 2 above)
Comment (bkirwi) [reply src/storage-types/src/dyncfgs.rs:25]: 🤦, will clean this up!

File: src/storage-controller/src/command_wals.rs (hunk 1 above)
Comment (bkirwi) [reply src/storage-controller/src/command_wals.rs:161]: Makes sense! Moved the check into the method so it can't be forgotten...

**Comment by bkirwi** [comment]
Going to get this in to avoid a bitrot, but if any other controller-interested folks have feedback I can take it in a followup!

**Pull request merged**